File: apps/android/app/src/main/java/com/voiid/app/onboarding/PhoneScreen.kt
package com.voiid.app.onboarding

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import kotlinx.coroutines.delay
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowLeft
import androidx.compose.material.icons.outlined.ChatBubbleOutline
import androidx.compose.material.icons.outlined.PersonOutline
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material.icons.outlined.People
import androidx.compose.material.icons.outlined.Shield
import androidx.compose.material.icons.outlined.VerifiedUser
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.components.softClickable
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont
import kotlinx.coroutines.launch

/**
 * Strip what autofill adds. Port of iOS `PhoneScreen.normalise`.
 *
 * THE GUARD IS THE LENGTH CHECK. A bare national number that happens to start with its own
 * dial digits — a Delhi landline starting "91…" — is only stripped when what remains is still
 * a plausible length, so a legitimate number is never mangled.
 *
 * Android had none of this: autofill returning "+91 98765 43210" was filtered to
 * "919876543210" and then prefixed with the dial code again, sending +91919876543210 — a
 * wrong number that looks correctly entered.
 */
internal fun normalisePhone(raw: String, country: Country): String {
    var d = raw.filter { it.isDigit() }
    val code = country.dialCode.removePrefix("+")

    if (d.length > country.maxDigits && d.startsWith(code)) {
        val stripped = d.removePrefix(code)
        if (stripped.length >= country.minDigits) d = stripped
    }

    // Some regions autofill a trunk "0" prefix ("098765..."). Same guard.
    if (d.length > country.maxDigits && d.startsWith("0")) {
        val stripped = d.removePrefix("0")
        if (stripped.length >= country.minDigits) d = stripped
    }

    return d.take(country.maxDigits)
}



/**
 * Onboarding — phone entry, built to the brand reference. Twin of iOS
 * `Onboarding/PhoneScreen.swift`; the two must stay identical.
 *
 * The Firebase send is unchanged from the previous version of this screen — only the
 * presentation is new.
 */
@Composable
fun PhoneScreen(
    onBack: () -> Unit,
    onContinue: (phone: String, verificationId: String) -> Unit,
) {
    val haptics = LocalVoiidHaptics.current
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    var phone by remember { mutableStateOf("") }
    var country by remember { mutableStateOf(CountryStore.default) }
    var showPicker by remember { mutableStateOf(false) }
    var sending by remember { mutableStateOf(false) }
    var errorText by remember { mutableStateOf<String?>(null) }
    var appeared by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) { appeared = true }

    val focus = LocalFocusManager.current
    val fieldFocus = remember { FocusRequester() }
    // 350ms, matching iOS. The field must be attached before focus is requested, and a
    // keyboard that appears before the screen has settled reads as a jump.
    LaunchedEffect(Unit) {
        delay(350)
        runCatching { fieldFocus.requestFocus() }
    }

    val interaction = remember { MutableInteractionSource() }
    val focused by interaction.collectIsFocusedAsState()

    /**
     * Digits only, normalised. Formatting characters a keyboard might insert never reach the
     * wire, and a pasted or autofilled number is stripped of its dial code and trunk zero.
     */
    val digits = normalisePhone(phone, country)

    /**
     * Plausible LENGTH for this country, not a validity check — several countries have
     * genuinely variable formats, and the SMS is the real validator.
     *
     * This was `digits.length >= 6` under a comment arguing that per-country lengths reject
     * legitimate numbers. iOS ships a 237-entry table doing exactly that, so the loose rule
     * was not a shared decision: it let a 6-digit Indian number through to Firebase and put
     * no ceiling on the other end at all.
     */
    val valid = digits.length >= country.minDigits && digits.length <= country.maxDigits

    // Send the OTP via Firebase, then advance to the OTP screen with the verificationId.
    // (Firebase texts the code; we verify it on the next screen.)
    fun sendOtp() {
        if (sending || !valid) return
        val activity = context as? android.app.Activity ?: run {
            errorText = "Can't start verification"; return
        }
        sending = true; errorText = null
        scope.launch {
            try {
                val e164 = "${country.dialCode}$digits"
                val verificationId = com.voiid.app.net.FirebasePhoneAuth.sendCode(activity, e164)
                haptics.tap(); onContinue(e164, verificationId)
            } catch (e: Exception) {
                errorText = e.message ?: "Couldn't send code"
                haptics.error()
            }
            sending = false
        }
    }

    Column(Modifier.fillMaxSize().background(VoiidBrand.ground).statusBarsPadding().imePadding()) {
        Row(Modifier.fillMaxWidth().height(48.dp).padding(horizontal = 12.dp), verticalAlignment = Alignment.CenterVertically) {
            androidx.compose.material3.IconButton(onClick = { focus.clearFocus(); haptics.tap(); onBack() }) {
                Icon(androidx.compose.material.icons.Icons.AutoMirrored.Filled.KeyboardArrowLeft,
                    contentDescription = "Back", tint = VoiidBrand.text, modifier = Modifier.size(28.dp))
            }
        }
        Column(Modifier.weight(1f).fillMaxWidth().verticalScroll(rememberScrollState()).padding(horizontal = 24.dp),
            horizontalAlignment = Alignment.CenterHorizontally) {
            OnboardingHeader(
                title = OnboardingTitleSpec.Stacked("Enter your", "phone number"),
                blurb = "We'll send you a verification code\nto confirm your number.",
            )
            Spacer(Modifier.height(24.dp))
            val borderColor by animateColorAsState(
                if (focused) VoiidBrand.lime
                else VoiidBrand.fieldEdge,
                tween(180), label = "phoneFieldBorder",
            )
            Row(
                Modifier
                    .fillMaxWidth()
                    .height(62.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .background(VoiidBrand.field)
                    .border(if (focused) 1.5.dp else 1.dp, borderColor, RoundedCornerShape(16.dp)),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Row(
                    Modifier
                        .height(62.dp)
                        .clip(RoundedCornerShape(14.dp))
                        .softClickable { haptics.tap(); focus.clearFocus(); showPicker = true }
                        .padding(horizontal = 16.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    Text(country.flag, fontSize = 22.sp)
                    Text(country.dialCode, style = VoiidFont.rounded(17, FontWeight.Medium),
                         color = VoiidBrand.text)
                    Icon(Icons.Default.KeyboardArrowDown, contentDescription = null,
                         tint = VoiidBrand.textDim, modifier = Modifier.size(12.dp))
                }

                Box(
                    Modifier
                        .width(1.dp)
                        .height(34.dp)
                        .background(VoiidBrand.hairline),
                )

                BasicTextField(
                    value = phone,
                    onValueChange = { raw ->
                        phone = normalisePhone(raw, country)
                        // Reaching the country's full length is the end of the task: the
                        // keyboard steps out of the way of Continue rather than sitting over
                        // it, and the soft press confirms the number is complete.
                        if (normalisePhone(raw, country).length == country.maxDigits) {
                            focus.clearFocus()
                            haptics.soft()
                        }
                    },
                    singleLine = true,
                    textStyle = VoiidFont.rounded(17, FontWeight.Medium).copy(color = VoiidBrand.text),
                    cursorBrush = SolidColor(VoiidBrand.lime),
                    interactionSource = interaction,
                    keyboardOptions = KeyboardOptions(
                        keyboardType = KeyboardType.Phone,
                        imeAction = ImeAction.Go,
                    ),
                    keyboardActions = KeyboardActions(onGo = { sendOtp() }),
                    modifier = Modifier.weight(1f).padding(horizontal = 16.dp)
                        .focusRequester(fieldFocus),
                    decorationBox = { inner ->
                        if (phone.isEmpty()) {
                            Text("Phone number", style = VoiidFont.rounded(17, FontWeight.Medium),
                                 color = VoiidBrand.placeholder)
                        }
                        inner()
                    },
                )
            }

            errorText?.let {
                Text(it, style = VoiidFont.rounded(13), color = VoiidColor.error,
                    textAlign = TextAlign.Center, modifier = Modifier.padding(top = 8.dp))
            }
            Spacer(Modifier.height(24.dp))
            PhonePromiseCard(Icons.Outlined.Lock, "Secure & private", "Your number is encrypted and always private.")
            Spacer(Modifier.height(10.dp))
            PhonePromiseCard(Icons.Outlined.ChatBubbleOutline, "No spam. Ever.", "We never share your number with anyone.")
            Spacer(Modifier.height(10.dp))
            PhonePromiseCard(Icons.Outlined.PersonOutline, "Used only for you", "To verify your identity and keep your account secure.")
            Spacer(Modifier.height(24.dp))
        }
        OnboardingFooter {
            OnboardingKitButton(title = "Continue", enabled = valid && !sending, busy = sending, usesBrandGradient = true) {
                focus.clearFocus()
                sendOtp()
            }
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.Center, verticalAlignment = Alignment.CenterVertically) {
                Icon(Icons.Outlined.Lock, null, tint = VoiidBrand.lime, modifier = Modifier.size(13.dp))
                Spacer(Modifier.width(8.dp))
                Text("We'll send a verification code by SMS.\nMessage and data rates may apply.",
                    style = VoiidFont.rounded(12.5f), color = VoiidBrand.textDim, textAlign = TextAlign.Center)
            }
        }
    }

    if (showPicker) {
        CountryPickerSheet(
            selected = country,
            onSelect = { picked ->
                // ONLY when it actually changed. A number typed for India is not a number
                // for Germany, so it is cleared rather than silently re-prefixed — and the
                // field takes focus back so the user can start typing immediately.
                val changed = picked.id != country.id
                country = picked
                if (changed) {
                    phone = ""
                    scope.launch { delay(50); runCatching { fieldFocus.requestFocus() } }
                }
            },
            onDismiss = { showPicker = false },
        )
    }
}

@Composable
private fun PhonePromiseCard(icon: androidx.compose.ui.graphics.vector.ImageVector, title: String, detail: String) {
    val shape = RoundedCornerShape(16.dp)
    Row(Modifier.fillMaxWidth().background(VoiidBrand.card, shape).border(1.dp, VoiidBrand.hairline, shape)
        .padding(horizontal = 16.dp, vertical = 14.dp),
        horizontalArrangement = Arrangement.spacedBy(16.dp), verticalAlignment = Alignment.CenterVertically) {
        Box(Modifier.size(46.dp).background(VoiidBrand.lime.copy(alpha = 0.1f), androidx.compose.foundation.shape.CircleShape),
            contentAlignment = Alignment.Center) {
            Icon(icon, null, tint = VoiidBrand.lime, modifier = Modifier.size(19.dp))
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(3.dp)) {
            Text(title, style = VoiidFont.rounded(15, FontWeight.SemiBold), color = VoiidBrand.text)
            Text(detail, style = VoiidFont.rounded(12.5f), color = VoiidBrand.textDim)
        }
        Box(Modifier.width(1.dp).height(40.dp).background(VoiidBrand.hairline))
        Icon(Icons.Outlined.VerifiedUser, null, tint = VoiidBrand.lime, modifier = Modifier.size(20.dp))
    }
}

File: apps/ios/Voiid/Voiid/Onboarding/PhoneScreen.swift
//
//  PhoneScreen.swift
//  Voiid
//
//  Enter your phone number — the onboarding step after Permissions.
//
//  Built to the design source (`Voiid Ui/Screens/PhoneNumberScreen.swift`) through
//  `OnboardingKit`, so the sizes, paddings and weights are the reference's own numbers. The
//  header is the WORDMARK at 34pt, matching Terms and Permissions.
//
//  ── THIS SCREEN DOES SEND ───────────────────────────────────────────────────────
//  The design source sends nothing — its `onContinue` just hands the number up. The real
//  Firebase send lives here and is UNCHANGED by the restyle: `FirebasePhoneAuth.sendCode`
//  returns a verificationID which the OTP step needs.
//
//  ── THE VALIDATION IS DELIBERATELY SHALLOW ──────────────────────────────────────
//  Continue enables on a plausible LENGTH for the chosen country, and nothing more. It does not
//  claim the number exists or that the user owns it — only the SMS can establish that, so
//  pretending otherwise here would be a lie the next screen has to walk back.
//
//  What it DOES prevent is the obvious dead end: tapping Continue on an empty or half-typed
//  field, waiting for an SMS, and getting nothing.
//
//  ── AUTOFILL HANDS BACK A FULL INTERNATIONAL NUMBER ─────────────────────────────
//  iOS autofill and Contacts supply "+91 98765 43210", not "9876543210". Keeping only the digits
//  leaves the dial code attached, which overflows the field and silently truncates the REAL
//  number — a wrong number that looks correctly entered, invisible until the SMS never arrives.
//  `normalise` strips it, guarded by a length check so a national number beginning with the same
//  digits survives.
//
//  ── THE KEYBOARD GETS OUT OF THE WAY ────────────────────────────────────────────
//  A number pad has no Return key, so it never dismisses itself — and it covers Continue. The
//  field drops focus the moment the number reaches its country's maximum length, which is the
//  one point where there is provably nothing left to type. Variable-length countries never trip
//  that, so there is also a keyboard-toolbar Done.
//
//  Committed to dark, like the rest of onboarding.
//

import SwiftUI

struct PhoneScreen: View {
    /// Passes the E.164 number + Firebase verificationID to the OTP step.
    let onContinue: (String, String) -> Void
    /// Back to the permissions step. Optional so the screen still works if it is ever the root.
    var onBack: (() -> Void)? = nil

    @State private var country = Country.default   // India default
    @State private var digits = ""
    @State private var showingCountries = false
    @State private var sending = false
    @State private var errorText: String?

    /// Drives the field's focused ring. The reference shows the field lit, which reads as "this
    /// is where you are" — so the screen opens with it focused rather than waiting for a tap.
    @FocusState private var fieldFocused: Bool

    /// Digits only, with the country's own dial code stripped, capped at its maximum length.
    ///
    /// The guard is the LENGTH check: a bare national number that happens to start with the dial
    /// digits (a Delhi landline starting "91…") is only stripped if what remains is still a
    /// plausible length, so a legitimate number is never mangled.
    private func normalise(_ raw: String) -> String {
        var d = raw.filter(\.isNumber)

        // The dial code without its "+", e.g. "91".
        let code = country.dialCode.dropFirst()

        if d.count > country.maxDigits, d.hasPrefix(code) {
            let stripped = String(d.dropFirst(code.count))
            if stripped.count >= country.minDigits { d = stripped }
        }

        // Some regions autofill a trunk "0" prefix ("098765..."). Same guard.
        if d.count > country.maxDigits, d.hasPrefix("0") {
            let stripped = String(d.dropFirst())
            if stripped.count >= country.minDigits { d = stripped }
        }

        return String(d.prefix(country.maxDigits))
    }

    /// Plausible length only. See the header on why this is not stricter.
    private var isPlausible: Bool {
        digits.count >= country.minDigits && digits.count <= country.maxDigits
    }

    /// Send the OTP via Firebase, then advance to the OTP screen with the verificationID.
    private func sendOtp() {
        guard !sending, isPlausible else { return }
        sending = true
        errorText = nil
        fieldFocused = false
        let e164 = "\(country.dialCode)\(digits)"
        Task {
            do {
                let verificationID = try await FirebasePhoneAuth.sendCode(to: e164)
                Haptics.tap()
                onContinue(e164, verificationID)
            } catch {
                errorText = error.localizedDescription
                Haptics.error()
            }
            sending = false
        }
    }

    var body: some View {
        ZStack {
            VoiidBrand.ground.ignoresSafeArea()

            ScrollView {
                VStack(spacing: 0) {
                    OnboardingHeader(
                        title: .stacked("Enter your", accent: "phone number"),
                        blurb: "We'll send you a verification code\nto confirm your number."
                    )

                    numberField
                        .padding(.top, VoiidSpacing.lg)

                    if let errorText {
                        Text(errorText)
                            .font(VoiidFont.rounded(13, .regular))
                            .foregroundColor(VoiidColor.error)
                            .multilineTextAlignment(.center)
                            .fixedSize(horizontal: false, vertical: true)
                            .padding(.top, VoiidSpacing.sm)
                    }

                    VStack(spacing: 10) {
                        ForEach(PhonePromise.all) { promise in
                            PromiseCard(promise: promise)
                        }
                    }
                    .padding(.top, VoiidSpacing.lg)
                }
                .padding(.horizontal, VoiidSpacing.lg)
                .padding(.bottom, 240)
            }
            .scrollIndicators(.hidden)
            // Tapping the background dismisses the keyboard. Without it the only way out of the
            // field on a screen with no other tap target is the Return key, which a number pad
            // does not have.
            .onTapGesture { fieldFocused = false }

            VStack(spacing: 0) {
                Spacer(minLength: 0)
                footer
            }
            .ignoresSafeArea(edges: .bottom)
        }
        .preferredColorScheme(.dark)
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(.hidden, for: .navigationBar)
        .sheet(isPresented: $showingCountries) {
            CountryPickerSheet(selected: $country) {
                // Clear on change: a number valid for one country is rarely valid for another,
                // and silently keeping it produces a number that looks entered but cannot pass.
                digits = ""
                // Straight back to typing — choosing a country is a step ON THE WAY to entering
                // a number, never the goal, so the field takes focus again without a tap.
                fieldFocused = true
            }
        }
        // Focused after a beat, not immediately: during a navigation PUSH the field is not yet
        // in the responder chain when `onAppear` fires, so setting focus there silently does
        // nothing — which leaves the field grey and the keyboard down. A hop to the next runloop
        // lands after the push completes.
        .task {
            try? await Task.sleep(for: .milliseconds(350))
            fieldFocused = true
        }
    }

    // MARK: The field

    private var numberField: some View {
        HStack(spacing: 0) {
            countryPicker

            // A hairline, not a gap: the dial code and the number are ONE value, and separating
            // them into two boxes would suggest they are entered independently.
            Rectangle()
                .fill(VoiidBrand.hairline)
                .frame(width: 1, height: 34)

            TextField("", text: $digits, prompt:
                Text("Phone number").foregroundColor(VoiidBrand.placeholder)
            )
            .keyboardType(.phonePad)          // includes +*# , which .numberPad omits — pasted
                                              // international numbers would otherwise be rejected
                                              // character-by-character before `normalise` sees them
            .textContentType(.telephoneNumber)
            .font(VoiidFont.rounded(17, .medium))
            .foregroundColor(VoiidBrand.text)
            .focused($fieldFocused)
            .padding(.horizontal, VoiidSpacing.md)
            // The escape hatch for VARIABLE-length countries. Auto-dismiss fires at maxDigits,
            // which a fixed-length country reaches on its last digit — but Germany's range is
            // 6–11, so a valid 9-digit number never trips it and the keyboard would stay up
            // over Continue forever. A toolbar Done is the standard way out of a number pad.
            .toolbar {
                ToolbarItemGroup(placement: .keyboard) {
                    Spacer()
                    Button("Done") { fieldFocused = false }
                        .font(VoiidFont.rounded(16, .semibold))
                        .foregroundColor(VoiidBrand.lime)
                }
            }
            // Digits only, capped at the country's own maximum. Filtering on change rather than
            // validating on submit: a keypad can still deliver paste, and a number that silently
            // exceeds its own format is worse than one that stops accepting characters.
            .onChange(of: digits) { _, new in
                let filtered = normalise(new)
                if filtered != new {
                    digits = filtered
                    return          // this fires again with the clean value; act on that pass
                }

                // ── THE KEYBOARD LEAVES WHEN THE NUMBER IS DONE ─────────────────────────
                // A number pad has no Return key, so nothing about it says "finished". The
                // keyboard therefore sat over Continue, and a user who had typed a complete
                // number had to know to tap the background to reach the button they were being
                // asked to press.
                //
                // Only at MAX, never at min. Countries whose range is wide (Germany is 6–11)
                // would otherwise have the keyboard yanked away mid-number, which is far worse
                // than leaving it up.
                if filtered.count == country.maxDigits {
                    fieldFocused = false
                    // Confirms the field took the last digit, on a screen where the keyboard
                    // vanishing is the only other signal.
                    Haptics.soft()
                }
            }
        }
        .frame(height: 62)
        .background(
            RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous)
                .fill(VoiidBrand.field)
        )
        .overlay(
            RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous)
                .stroke(
                    fieldFocused ? VoiidBrand.lime : VoiidBrand.fieldEdge,
                    lineWidth: fieldFocused ? 1.5 : 1
                )
        )
        .animation(.easeOut(duration: 0.18), value: fieldFocused)
    }

    private var countryPicker: some View {
        Button {
            Haptics.tap()
            fieldFocused = false      // the keyboard would otherwise fight the sheet
            showingCountries = true
        } label: {
            HStack(spacing: 8) {
                Text(country.flag)
                    .font(.system(size: 22))
                Text(country.dialCode)
                    .font(VoiidFont.rounded(17, .medium))
                    .foregroundColor(VoiidBrand.text)
                Image(systemName: "chevron.down")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundColor(VoiidBrand.textDim)
            }
            .padding(.horizontal, VoiidSpacing.md)
            .frame(maxHeight: .infinity)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Country code, \(country.name), \(country.dialCode)")
        .accessibilityHint("Opens the country list")
    }

    // MARK: Footer

    private var footer: some View {
        OnboardingFooter {
            OnboardingKitButton(title: "Continue", enabled: isPlausible && !sending) {
                sendOtp()
            }

            // What Continue will actually do. Centred under the button, matching the rest of
            // the screen's centred type.
            HStack(spacing: 8) {
                Image(systemName: "lock")
                    .font(.system(size: 13, weight: .medium))
                    .foregroundColor(VoiidBrand.lime)

                Text("We'll send a verification code by SMS.\nMessage and data rates may apply.")
                    .font(VoiidFont.rounded(12.5))
                    .foregroundColor(VoiidBrand.textDim)
                    .multilineTextAlignment(.center)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .frame(maxWidth: .infinity)
        }
    }
}

// MARK: - A promise card

/// One reassurance: icon in a lime disc, copy, and a checkmark holding the right edge.
///
/// The trailing check is separated by a hairline rather than floating: without it the tick reads
/// as an interactive control the user is meant to tap, which is exactly what it is not.
private struct PromiseCard: View {
    let promise: PhonePromise

    var body: some View {
        HStack(spacing: VoiidSpacing.md) {
            Circle()
                .fill(VoiidBrand.lime.opacity(0.10))
                .frame(width: 46, height: 46)
                .overlay {
                    Image(systemName: promise.icon)
                        .font(.system(size: 19, weight: .medium))
                        .foregroundColor(VoiidBrand.lime)
                }

            VStack(alignment: .leading, spacing: 3) {
                Text(promise.title)
                    .font(VoiidFont.rounded(15, .semibold))
                    .foregroundColor(VoiidBrand.text)

                Text(promise.detail)
                    .font(VoiidFont.rounded(12.5))
                    .foregroundColor(VoiidBrand.textDim)
                    .fixedSize(horizontal: false, vertical: true)
                    .multilineTextAlignment(.leading)
            }

            Spacer(minLength: 0)

            Rectangle()
                .fill(VoiidBrand.hairline)
                .frame(width: 1, height: 40)

            Image(systemName: "checkmark.shield")
                .font(.system(size: 20, weight: .medium))
                .foregroundColor(VoiidBrand.lime)
                .padding(.leading, 4)
        }
        .padding(.horizontal, VoiidSpacing.md)
        .padding(.vertical, 14)
        .background(VoiidBrand.card)
        .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous)
                .stroke(VoiidBrand.hairline, lineWidth: 1)
        )
        .accessibilityElement(children: .combine)
    }
}

// MARK: - Data

/// A reassurance shown under the field.
struct PhonePromise: Identifiable, Hashable {
    let id: String
    let title: String
    let detail: String
    let icon: String

    static let all: [PhonePromise] = [
        .init(id: "secure",
              title: "Secure & private",
              detail: "Your number is encrypted and always private.",
              icon: "lock"),
        .init(id: "nospam",
              title: "No spam. Ever.",
              detail: "We never share your number with anyone.",
              icon: "ellipsis.bubble"),
        .init(id: "identity",
              title: "Used only for you",
              detail: "To verify your identity and keep your account secure.",
              icon: "person"),
    ]
}
